src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt

```diff
@@ -120,53 +121,56 @@ class VisitService(
     return bookedVisitDto
   }
 
-  fun getBookCountForSlot(sessionSlotId: Long, restriction: VisitRestriction): Long {
-    return if (VisitRestriction.OPEN == restriction) {
-      visitRepository.getCountOfBookedForOpenSessionSlot(sessionSlotId)
-    } else {
-      visitRepository.getCountOfBookedForClosedSessionSlot(sessionSlotId)
+  private fun checkSlotCapacity(
+    bookingRequestDto: BookingRequestDto,
+    application: Application,
+    existingBooking: Visit?,
+  ) {
+    val differentSlotFromLastBooking = applicationRequiresANewSlot(existingBooking, application)
+
+    if (!bookingRequestDto.allowOverBooking && differentSlotFromLastBooking) {
+      slotCapacityService.checkCapacityForBooking(
+        application.sessionSlot.reference,
+        application.restriction,
+        applicationService.isExpiredApplication(application.modifyTimestamp!!),
+      )
     }
   }
 
-  private fun hasExceededMaxCapacity(
+  private fun applicationRequiresANewSlot(
+    existingBooking: Visit?,
     application: Application,
   ): Boolean {
-    var slotTakenCount = getBookCountForSlot(application.sessionSlotId, application.restriction)
-
-    if (isExpiredApplication(application.modifyTimestamp!!)) {
-      slotTakenCount += applicationService.getReservedApplicationsCountForSlot(application.sessionSlotId, application.restriction)
+    var differentSlotFromLastBooking = true
+    existingBooking?.let {
+      differentSlotFromLastBooking = existingBooking.visitRestriction != application.restriction || it.sessionSlot.id != application.sessionSlotId
     }
-
-    val maxBookings = sessionTemplateService.getSessionTemplatesCapacity(
-      application.sessionSlot.sessionTemplateReference!!,
-      application.restriction,
-    )
-
-    return slotTakenCount >= maxBookings
+    return differentSlotFromLastBooking
   }
 
-  private fun isExpiredApplication(modifyTimestamp: LocalDateTime): Boolean {
-    return modifyTimestamp.isBefore(applicationService.getExpiredApplicationDateAndTime())
-  }
-
-  private fun createBooking(application: Application, hasExistingBooking: Boolean): Visit {
-    val existingBooking = if (hasExistingBooking) visitRepository.findVisitByApplicationReference(application.reference) else null
-    if (hasExistingBooking) {
-      validateVisitStartDate(existingBooking!!, "changed")
-      handleVisitUpdateEvents(existingBooking, application)
+  fun getBookCountForSlot(sessionSlotId: Long, restriction: VisitRestriction): Long {
```

Comment:
should this be SessionRestriction so that only OPEN and CLOSED are considered and not UNKNOWN?

Reply:
updated